xdsl/parser.py

```diff
@@ -1,1335 +1,1924 @@
     """
-    The character index in the entire file.
-    A line break is consider to be a character here.
+    input: Input
     """
+    The input being operated on
+    """
+
+    def __len__(self):
+        return self.len
+
+    @property
+    def len(self):
+        return self.end - self.start
 
-    line: int = field(default=1)
-    """The line index."""
+    @property
+    def text(self):
+        return self.input.content[self.start:self.end]
 
-    column: int = field(default=1)
-    """The character index in the current line."""
+    def get_line_col(self) -> tuple[int, int]:
+        info = self.input.get_lines_containing(self)
+        if info is None:
+            return -1, -1
+        lines, offset_of_first_line, line_no = info
+        return line_no, self.start - offset_of_first_line
+
+    def print_with_context(self, msg: str | None = None) -> str:
+        """
+        returns a string containing lines relevant to the span. The Span's contents
+        are highlighted by up-carets beneath them (`^`). The message msg is printed
+        along these.
+        """
+        info = self.input.get_lines_containing(self)
+        if info is None:
+            return "Unknown location of span {}. Error: ".format(self, msg)
+        lines, offset_of_first_line, line_no = info
+        # offset relative to the first line:
+        offset = self.start - offset_of_first_line
+        remaining_len = max(self.len, 1)
+        capture = StringIO()
+        print("{}:{}:{}".format(self.input.name, line_no, offset),
+              file=capture)
+        for line in lines:
+            print(line, file=capture)
+            if remaining_len < 0:
+                continue
+            len_on_this_line = min(remaining_len, len(line) - offset)
+            remaining_len -= len_on_this_line
+            print("{}{}".format(" " * offset, "^" * max(len_on_this_line, 1)),
+                  file=capture)
+            if msg is not None:
+                print("{}{}".format(" " * offset, msg), file=capture)
+                msg = None
+            offset = 0
+        if msg is not None:
+            print(msg, file=capture)
+        return capture.getvalue()
+
+    def __repr__(self):
+        return "{}[{}:{}](text='{}')".format(self.__class__.__name__,
+                                             self.start, self.end, self.text)
+
+
+@dataclass(frozen=True, repr=False)
+class StringLiteral(Span):
 
-    def __str__(self):
-        return f"{self.line}:{self.column}"
+    def __post_init__(self):
+        if len(self) < 2 or self.text[0] != '"' or self.text[-1] != '"':
+            raise ParseError(self, "Invalid string literal!")
 
-    def next_char_pos(self, n: int = 1) -> Position | None:
-        """Return the position of the next character in the string."""
-        if self.idx >= len(self.file) - n:
+    @classmethod
+    def from_span(cls, span: Span | None) -> StringLiteral | None:
```

Comment:
This makes it easier to use imo, because you can just do 

```python
return StringLiteral.from_span(self.try_parse_string_literal())
```

instead of

```python
span = self.try_parse_string_literal()
if span is None:
    return None
return StrngLiteral.from_span(span)
```

Reply:
Okay makes sense!